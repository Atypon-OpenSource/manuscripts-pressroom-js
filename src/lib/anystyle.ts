/*!
 * Â© 2020 Atypon Systems LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { CSL } from '@manuscripts/manuscript-transform'
import { execFile } from 'child_process'
import { promisify } from 'util'

export const parseReferences = async (
  inputPath: string
): Promise<CSL.Item[]> => {
  const { stdout } = await promisify(execFile)('anystyle', [
    '--stdout',
    '--format=csl',
    'parse',
    inputPath,
  ])

  return JSON.parse(stdout)
}
