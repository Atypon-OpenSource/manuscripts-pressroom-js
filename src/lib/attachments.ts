/*!
 * Â© 2022 Atypon Systems LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { ContainedModel } from '@manuscripts/manuscript-transform'
import {
  Figure,
  FigureElement,
  ObjectTypes,
} from '@manuscripts/manuscripts-json-schema'
import createHttpError from 'http-errors'

import { processElements, XLINK_NAMESPACE } from './data'
import { logger } from './logger'

export interface BasicAttachmentData {
  url: string
  name: string
}

export interface AttachmentData {
  url: string
  name: string
  designation: string
  description: string
  MIME: string
}

export type FiguresAndAttachmentsData = {
  figuresMap: Map<string, FigureElement>
}

export const generateFiguresMap = (
  data: Array<ContainedModel>
): Map<string, FigureElement> => {
  const figures = data.filter(
    (el) => el.objectType == ObjectTypes.FigureElement
  ) as Array<FigureElement>
  const figuresMap = new Map<string, FigureElement>()
  for (const figure of figures) {
    figuresMap.set(figure._id, figure)
  }
  return figuresMap
}

export const generateGraphicsMap = (
  data: Array<ContainedModel>
): Map<string, Figure> => {
  const graphics = data.filter(
    (el) => el.objectType == ObjectTypes.Figure
  ) as Array<Figure>
  const graphicsMap = new Map<string, Figure>()
  for (const graphic of graphics) {
    graphicsMap.set(graphic._id, graphic)
  }
  return graphicsMap
}

export const generateAttachmentsMap = (
  attachments: Array<AttachmentData>
): Map<string, AttachmentData> => {
  const attachmentsMap = new Map<string, AttachmentData>()
  for (const attachment of attachments) {
    attachmentsMap.set(attachment.url, attachment)
  }

  return attachmentsMap
}

export const generateBasicAttachmentsMap = (
  attachments: BasicAttachmentData[]
): Map<string, BasicAttachmentData> => {
  const attachmentsMap = new Map<string, BasicAttachmentData>()
  attachments.forEach((a) => attachmentsMap.set(a.url, a))
  return attachmentsMap
}

export const exportAttachments = async (
  document: Document,
  figures: Map<string, FigureElement>,
  graphics: Map<string, Figure>,
  attachmentsMap: Map<string, AttachmentData>,
  supplementaryDOI: Map<string, string>
): Promise<Document> => {
  const articleMeta = document.querySelector('article-meta')
  for (const [, figure] of figures) {
    const graphicIDs = figure.containedObjectIDs.filter((i) =>
      i.startsWith('MPFigure')
    )
    for (const graphicID of graphicIDs) {
      await processElements(
        document,
        `//graphic[starts-with(@xlink:href,"graphic/${graphicID.replace(
          ':',
          '_'
        )}")]`,
        async (graphic) => {
          const url = findImageRepresentation(graphics.get(graphicID))
          if (url) {
            const attachment = attachmentsMap.get(url)
            if (!attachment) {
              return
            }

            const imageName = attachment.name
            graphic.setAttributeNS(XLINK_NAMESPACE, 'href', imageName)
          }

          const parentFigure = graphic.closest('fig')
          if (!parentFigure) {
            logger.error('graphic not wrapped inside a fig element')
            return
          }

          if (figure.alternatives) {
            const interactiveHtml: Array<string> = []
            const downloadable: Array<string> = []
            figure.alternatives.forEach(({ src }) => {
              const attachment = attachmentsMap.get(src)
              if (!attachment) {
                return
              }
              if (attachment.designation === 'interactive-html') {
                interactiveHtml.push(src)
              } else {
                downloadable.push(src)
              }
            })

            if (downloadable.length > 0) {
              const caption = document.createElement('caption')
              caption.setAttribute('content-type', 'supplementary-material')
              downloadable.forEach((url) => {
                const attachment = attachmentsMap.get(url)
                if (attachment) {
                  const inlineSupplementary = createInlineSupplementaryMaterial(
                    attachment,
                    document
                  )
                  caption.appendChild(inlineSupplementary)
                }
                parentFigure.insertBefore(caption, graphic)
              })
            }

            if (interactiveHtml.length > 0) {
              const alternatives = document.createElement('alternatives')
              const clone = graphic.cloneNode(true)
              alternatives.appendChild(clone)
              interactiveHtml.forEach((url) => {
                const attachment = attachmentsMap.get(url)
                if (attachment) {
                  const supplementary = createSupplementaryMaterial(
                    attachment,
                    document
                  )
                  const objectID = supplementaryDOI.get(url)
                  if (objectID) {
                    const objectIDNode = buildObjectID(objectID)
                    supplementary.appendChild(objectIDNode)
                  } else {
                    throw createHttpError(400, `DOI for ${url} not found`)
                  }
                  if (articleMeta) {
                    articleMeta.appendChild(supplementary.cloneNode(true))
                  }
                  alternatives.appendChild(supplementary)
                }
              })
              graphic.replaceWith(alternatives)
            }
          }
        }
      )
    }
  }

  return document
}

export const replaceHTMLImgReferences = async (
  document: Document,
  graphics: Map<string, Figure>,
  attachmentsMap: Map<string, AttachmentData>
): Promise<Document> => {
  for (const [, graphic] of graphics) {
    const name = graphic._id.replace(':', '_')
    await processElements(
      document,
      `//img[contains(@src,"${name}")]`,
      async (graphic: Element) => {
        const src = graphic.getAttribute('src')
        if (src) {
          const staticImage = attachmentsMap.get(src)
          if (staticImage) {
            const imageName = staticImage.name
            graphic.setAttribute('src', imageName)
          }
        }
      }
    )
  }
  return document
}

function buildObjectID(doi: string) {
  const objectID = new DOMParser().parseFromString(
    '<object-id pub-id-type="doi" specific-use="metadata"></object-id>',
    'application/xml'
  ).firstElementChild
  if (objectID) {
    objectID.textContent = doi
    return objectID
  }
  throw new Error('Unable to create object id for ' + doi)
}

const createInlineSupplementaryMaterial = (
  attachment: AttachmentData,
  document: Document
): Element => {
  const supplementaryMaterial = document.createElement(
    'inline-supplementary-material'
  )
  setSupplementaryMaterialAttributes(supplementaryMaterial, attachment)
  const rootParagraph = document.createElement('p')

  rootParagraph.appendChild(supplementaryMaterial)
  return rootParagraph
}

const createSupplementaryMaterial = (
  attachment: AttachmentData,
  document: Document
): Element => {
  const supplementaryMaterial = document.createElement('supplementary-material')

  setSupplementaryMaterialAttributes(supplementaryMaterial, attachment)
  return supplementaryMaterial
}

const setSupplementaryMaterialAttributes = (
  supplementaryMaterial: Element,
  attachment: AttachmentData
) => {
  supplementaryMaterial.setAttribute('mimetype', attachment.MIME)
  supplementaryMaterial.setAttributeNS(
    XLINK_NAMESPACE,
    'href',
    `external/${attachment.name}`
  )
  supplementaryMaterial.setAttribute('content-type', attachment.designation)
  if (attachment.description) {
    supplementaryMaterial.textContent = attachment.description
  }
}

export const findImageRepresentation = (
  graphic: Figure | undefined
): string | undefined => (graphic ? graphic.src : undefined)
