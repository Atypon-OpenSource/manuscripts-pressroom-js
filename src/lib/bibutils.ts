/*!
 * Â© 2020 Atypon Systems LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { execFile } from 'child_process'
import { promisify } from 'util'

type OutputFormat = 'ads' | 'end' | 'isi' | 'ris' | 'wordbib'

const formats: Record<OutputFormat, string> = {
  ads: 'xml2ads',
  end: 'xml2end',
  isi: 'xml2isi',
  ris: 'xml2ris',
  wordbib: 'xml2wordbib',
}

export const convertBibtex = async (
  bibtex: Buffer,
  format: OutputFormat
): Promise<Buffer> => {
  if (!(format in formats)) {
    throw new Error('Unknown format')
  }

  // TODO: bib2xml?
  const convertToXML = promisify(execFile)('biblatex2xml', {
    encoding: 'buffer',
  })

  if (!convertToXML.child.stdin) {
    throw new Error('No stdin')
  }

  convertToXML.child.stdin.write(bibtex)
  convertToXML.child.stdin.end()

  const { stdout: xml } = await convertToXML

  const convertToFormat = promisify(execFile)(formats[format], {
    encoding: 'buffer',
  })

  if (!convertToFormat.child.stdin) {
    throw new Error('No stdin')
  }

  convertToFormat.child.stdin.write(xml)
  convertToFormat.child.stdin.end()

  const { stdout: output } = await convertToFormat

  return output
}
