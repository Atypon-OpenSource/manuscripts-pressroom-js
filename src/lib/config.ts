/*!
 * Â© 2020 Atypon Systems LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
const environmentVariable = (name: string): string => {
  const value = process.env[name]

  if (value === undefined || value === '') {
    if (process.env.NODE_ENV === 'test') {
      return ''
    } else {
      throw new Error(`Missing environment variable ${name}`)
    }
  }

  return value
}

const booleanEnvironmentVariable = (name: string) => {
  const value = environmentVariable(name)

  if (!value || value === '0' || value === 'false') {
    return false
  }

  return true
}

interface Config {
  api_key: string
  authorization: {
    emails: string
  }
  jwt: {
    enabled: boolean
    issuer: string
    root: string
  }
}

export const config: Config = {
  api_key: environmentVariable('PRESSROOM_API_KEY'),
  authorization: {
    emails: environmentVariable('PRESSROOM_AUTHORIZATION_EMAILS'),
  },
  jwt: {
    enabled: booleanEnvironmentVariable('PRESSROOM_JWT_ENABLED'),
    issuer: environmentVariable('PRESSROOM_JWT_ISSUER'),
    root: environmentVariable('PRESSROOM_JWT_ROOT'),
  },
}
