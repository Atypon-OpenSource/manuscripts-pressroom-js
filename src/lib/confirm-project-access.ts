import { RequestHandler } from 'express'
import createError from 'http-errors'

export const confirmProjectAccess: RequestHandler = (req, res, next) => {
  if (req.user.containerID !== req.params.projectID) {
    throw createError(401)
  }

  next()
}
