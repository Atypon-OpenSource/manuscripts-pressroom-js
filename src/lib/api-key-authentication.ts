import { RequestHandler } from 'express'

import { config } from './config'

export const apiKeyAuthentication: RequestHandler = (req, res, next) => {
  if (req.headers['pressroom-api-key'] !== config.api_key) {
    throw new Error('Incorrect API key')
  }
  next()
}
