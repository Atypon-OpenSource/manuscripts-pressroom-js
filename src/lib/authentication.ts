/*!
 * Â© 2020 Atypon Systems LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { RequestHandler } from 'express'
import jwt from 'express-jwt'
import createHttpError from 'http-errors'
import jwksRsa from 'jwks-rsa'

import { config } from './config'

export const apiKeyAuthentication: RequestHandler = (req, res, next) => {
  if (req.headers['pressroom-api-key'] !== config.api_key) {
    throw createHttpError(401, 'Incorrect API key')
  }
  req.user = {}
  next()
}

export const jwtAuthentication = jwt({
  secret: jwksRsa.expressJwtSecret({
    cache: true,
    rateLimit: true,
    jwksRequestsPerMinute: 5,
    jwksUri: `${config.jwt.root}/.well-known/jwks.json`,
  }),
  issuer: config.jwt.issuer,
  audience: 'pressroom-js',
  algorithms: ['RS256'],
})

export const authentication: RequestHandler = (req, res, next) => {
  if (config.api_key && 'pressroom-api-key' in req.headers) {
    apiKeyAuthentication(req, res, next)
  } else {
    jwtAuthentication(req, res, next)
  }
}
