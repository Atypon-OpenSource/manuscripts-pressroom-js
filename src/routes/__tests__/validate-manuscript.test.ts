/*!
 * Â© 2020 Atypon Systems LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { AnyValidationResult } from '@manuscripts/requirements'
import request from 'supertest'

import { config } from '../../lib/config'

describe('validate manuscript', () => {
  test('validates a manuscript with a test template', async () => {
    const { app } = await import('../../app')

    const response = await request(app)
      .post('/api/v2/validate/manuscript')
      .set('Accept', 'application/json')
      .attach('file', __dirname + '/__fixtures__/manuscript.manuproj')
      .field(
        'manuscriptID',
        'MPManuscript:9E0BEDBC-1084-4AA1-AB82-10ACFAE02232'
      )
      .field(
        'templateID',
        'MPManuscriptTemplate:www-zotero-org-styles-nature-genetics-Nature-Genetics-Journal-Publication-Article'
      )
      .set('pressroom-api-key', config.api_key)
      .responseType('json')

    expect(response.status).toBe(200)
    expect(response.get('Content-Type')).toBe('application/json; charset=utf-8')
    const results = JSON.parse(
      response.body.toString()
    ) as Array<AnyValidationResult>
    results.forEach((result) => {
      expect(result).toMatchSnapshot(
        {
          _id: expect.any(String),
        },
        'validate-manuscript'
      )
    })
  })
})
