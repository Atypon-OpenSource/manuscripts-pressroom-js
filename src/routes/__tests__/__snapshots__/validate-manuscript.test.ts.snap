// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`validate manuscript validates a manuscript with a test template: validate-manuscript 1`] = `
Object {
  "_id": Any<String>,
  "data": Object {
    "sectionCategory": "MPSectionCategory:introduction",
    "sectionDescription": Object {
      "_id": "MPSectionDescription:30856582-BC9E-426A-B4C5-56F5F681CDF8",
      "objectType": "MPSectionDescription",
      "required": true,
      "sectionCategory": "MPSectionCategory:introduction",
    },
  },
  "fixable": true,
  "message": "There must be Introduction section",
  "objectType": "MPRequiredSectionValidationResult",
  "passed": false,
  "severity": 0,
  "type": "required-section",
}
`;

exports[`validate manuscript validates a manuscript with a test template: validate-manuscript 2`] = `
Object {
  "_id": Any<String>,
  "data": Object {
    "sectionCategory": "MPSectionCategory:materials-method",
    "sectionDescription": Object {
      "_id": "MPSectionDescription:03FC1135-1F04-49A0-9D8F-E595EC2E6863",
      "objectType": "MPSectionDescription",
      "required": true,
      "sectionCategory": "MPSectionCategory:materials-method",
      "title": "online methods",
    },
  },
  "fixable": true,
  "message": "There must be Materials & Methods section",
  "objectType": "MPRequiredSectionValidationResult",
  "passed": false,
  "severity": 0,
  "type": "required-section",
}
`;

exports[`validate manuscript validates a manuscript with a test template: validate-manuscript 3`] = `
Object {
  "_id": Any<String>,
  "data": Object {
    "sectionCategory": "MPSectionCategory:abstract",
    "sectionDescription": Object {
      "_id": "MPSectionDescription:9F8AC4A6-0B93-46A9-ADF4-BC8F6A6603BE",
      "maxParagraphsCount": 1,
      "maxTitleCharLength": 400,
      "maxWordCount": 400,
      "objectType": "MPSectionDescription",
      "required": true,
      "sectionCategory": "MPSectionCategory:abstract",
    },
  },
  "fixable": true,
  "message": "There must be Abstract section",
  "objectType": "MPRequiredSectionValidationResult",
  "passed": false,
  "severity": 0,
  "type": "required-section",
}
`;

exports[`validate manuscript validates a manuscript with a test template: validate-manuscript 4`] = `
Object {
  "_id": Any<String>,
  "data": Object {
    "sectionCategory": "MPSectionCategory:results",
    "sectionDescription": Object {
      "_id": "MPSectionDescription:6C7F0155-A41D-456E-B396-E06FEB081CA9",
      "objectType": "MPSectionDescription",
      "required": true,
      "sectionCategory": "MPSectionCategory:results",
      "title": "Results",
    },
  },
  "fixable": true,
  "message": "There must be Results section",
  "objectType": "MPRequiredSectionValidationResult",
  "passed": false,
  "severity": 0,
  "type": "required-section",
}
`;

exports[`validate manuscript validates a manuscript with a test template: validate-manuscript 5`] = `
Object {
  "_id": Any<String>,
  "data": Object {
    "sectionCategory": "MPSectionCategory:competing-interests",
    "sectionDescription": Object {
      "_id": "MPSectionDescription:C37FB688-B8E2-44A3-959A-C76CA1C405A3",
      "objectType": "MPSectionDescription",
      "required": true,
      "sectionCategory": "MPSectionCategory:competing-interests",
      "title": "Competing financial interest",
    },
  },
  "fixable": true,
  "message": "There must be Competing Interests section",
  "objectType": "MPRequiredSectionValidationResult",
  "passed": false,
  "severity": 0,
  "type": "required-section",
}
`;

exports[`validate manuscript validates a manuscript with a test template: validate-manuscript 6`] = `
Object {
  "_id": Any<String>,
  "data": Object {
    "sectionCategory": "MPSectionCategory:discussion",
    "sectionDescription": Object {
      "_id": "MPSectionDescription:43239655-328D-4A07-A587-B21BF8E4230F",
      "objectType": "MPSectionDescription",
      "required": true,
      "sectionCategory": "MPSectionCategory:discussion",
      "title": "Discussion",
    },
  },
  "fixable": true,
  "message": "There must be Discussion section",
  "objectType": "MPRequiredSectionValidationResult",
  "passed": false,
  "severity": 0,
  "type": "required-section",
}
`;

exports[`validate manuscript validates a manuscript with a test template: validate-manuscript 7`] = `
Object {
  "_id": Any<String>,
  "data": Object {
    "sectionCategory": "MPSectionCategory:keywords",
    "sectionDescription": Object {
      "_id": "MPSectionDescription:B16D5504-DB50-4FE4-B2EE-6B77223CE511",
      "maxKeywordCount": 6,
      "minKeywordCount": 4,
      "objectType": "MPSectionDescription",
      "required": true,
      "sectionCategory": "MPSectionCategory:keywords",
    },
  },
  "fixable": true,
  "message": "There must be Keywords section",
  "objectType": "MPRequiredSectionValidationResult",
  "passed": false,
  "severity": 0,
  "type": "required-section",
}
`;

exports[`validate manuscript validates a manuscript with a test template: validate-manuscript 8`] = `
Object {
  "_id": Any<String>,
  "data": Object {
    "count": 458,
    "value": 400,
  },
  "message": "The manuscript must have less than or equal to 400 characters",
  "objectType": "MPCountValidationResult",
  "passed": false,
  "severity": 0,
  "type": "manuscript-maximum-characters",
}
`;

exports[`validate manuscript validates a manuscript with a test template: validate-manuscript 9`] = `
Object {
  "_id": Any<String>,
  "data": Object {
    "count": 458,
    "value": 100,
  },
  "objectType": "MPCountValidationResult",
  "passed": true,
  "severity": 0,
  "type": "manuscript-minimum-characters",
}
`;

exports[`validate manuscript validates a manuscript with a test template: validate-manuscript 10`] = `
Object {
  "_id": Any<String>,
  "data": Object {
    "count": 80,
    "value": 4000,
  },
  "objectType": "MPCountValidationResult",
  "passed": true,
  "severity": 0,
  "type": "manuscript-maximum-words",
}
`;

exports[`validate manuscript validates a manuscript with a test template: validate-manuscript 11`] = `
Object {
  "_id": Any<String>,
  "data": Object {
    "count": 80,
    "value": 2000,
  },
  "message": "The manuscript must have more than or equal to 2000 words",
  "objectType": "MPCountValidationResult",
  "passed": false,
  "severity": 0,
  "type": "manuscript-minimum-words",
}
`;

exports[`validate manuscript validates a manuscript with a test template: validate-manuscript 12`] = `
Object {
  "_id": Any<String>,
  "data": Object {
    "id": "MPSection:FFC7095E-A048-4CC4-BFF3-ADA4AF2C919E",
    "sectionCategory": "MPSectionCategory:bibliography",
  },
  "message": "Bibliography section must contains content",
  "objectType": "MPSectionBodyValidationResult",
  "passed": false,
  "severity": 0,
  "type": "section-body-has-content",
}
`;

exports[`validate manuscript validates a manuscript with a test template: validate-manuscript 13`] = `
Object {
  "_id": Any<String>,
  "data": Object {
    "count": 3,
    "value": 50,
  },
  "objectType": "MPCountValidationResult",
  "passed": true,
  "severity": 0,
  "type": "manuscript-title-maximum-words",
}
`;

exports[`validate manuscript validates a manuscript with a test template: validate-manuscript 14`] = `
Object {
  "_id": Any<String>,
  "data": Object {
    "count": 3,
    "value": 10,
  },
  "message": "The manuscript title must have more than or equal to 10 words",
  "objectType": "MPCountValidationResult",
  "passed": false,
  "severity": 0,
  "type": "manuscript-title-minimum-words",
}
`;

exports[`validate manuscript validates a manuscript with a test template: validate-manuscript 15`] = `
Object {
  "_id": Any<String>,
  "data": Object {
    "count": 16,
    "value": 80,
  },
  "objectType": "MPCountValidationResult",
  "passed": true,
  "severity": 0,
  "type": "manuscript-title-maximum-characters",
}
`;

exports[`validate manuscript validates a manuscript with a test template: validate-manuscript 16`] = `
Object {
  "_id": Any<String>,
  "data": Object {
    "count": 16,
    "value": 20,
  },
  "message": "The manuscript title must have more than or equal to 20 characters",
  "objectType": "MPCountValidationResult",
  "passed": false,
  "severity": 0,
  "type": "manuscript-title-minimum-characters",
}
`;

exports[`validate manuscript validates a manuscript with a test template: validate-manuscript 17`] = `
Object {
  "_id": Any<String>,
  "data": Object {
    "count": 1,
    "value": 15,
  },
  "objectType": "MPCountValidationResult",
  "passed": true,
  "severity": 0,
  "type": "manuscript-maximum-references",
}
`;

exports[`validate manuscript validates a manuscript with a test template: validate-manuscript 18`] = `
Object {
  "_id": Any<String>,
  "affectedElementId": "MPBibliographyItem:6C6AF046-1C45-4CC5-9F38-93BC736E5293",
  "message": "DOI is required",
  "objectType": "MPBibliographyValidationResult",
  "passed": false,
  "severity": 0,
  "type": "bibliography-doi-exist",
}
`;

exports[`validate manuscript validates a manuscript with a test template: validate-manuscript 19`] = `
Object {
  "_id": Any<String>,
  "data": Object {
    "count": 1,
    "value": 5,
  },
  "objectType": "MPCountValidationResult",
  "passed": true,
  "severity": 0,
  "type": "manuscript-maximum-figures",
}
`;

exports[`validate manuscript validates a manuscript with a test template: validate-manuscript 20`] = `
Object {
  "_id": Any<String>,
  "data": Object {
    "count": 1,
    "value": 3,
  },
  "objectType": "MPCountValidationResult",
  "passed": true,
  "severity": 0,
  "type": "manuscript-maximum-tables",
}
`;

exports[`validate manuscript validates a manuscript with a test template: validate-manuscript 21`] = `
Object {
  "_id": Any<String>,
  "data": Object {
    "count": 2,
    "value": 3,
  },
  "objectType": "MPCountValidationResult",
  "passed": true,
  "severity": 0,
  "type": "manuscript-maximum-combined-figure-tables",
}
`;

exports[`validate manuscript validates a manuscript with a test template: validate-manuscript 22`] = `
Object {
  "_id": Any<String>,
  "data": Object {
    "contentType": "image/png",
    "id": "MPFigure:C9A3CBCE-4AF4-4AD2-97B0-D2848BEFDE5A",
  },
  "message": "image/png format is not allowed",
  "objectType": "MPFigureFormatValidationResult",
  "passed": false,
  "severity": 0,
  "type": "figure-format-validation",
}
`;

exports[`validate manuscript validates a manuscript with a test template: validate-manuscript 23`] = `
Object {
  "_id": Any<String>,
  "affectedElementId": "MPFigure:C9A3CBCE-4AF4-4AD2-97B0-D2848BEFDE5A",
  "objectType": "MPFigureImageValidationResult",
  "passed": true,
  "severity": 0,
  "type": "figure-contains-image",
}
`;

exports[`validate manuscript validates a manuscript with a test template: validate-manuscript 24`] = `
Object {
  "_id": Any<String>,
  "data": Object {
    "count": 304,
    "id": "MPFigure:C9A3CBCE-4AF4-4AD2-97B0-D2848BEFDE5A",
    "value": 600,
  },
  "message": "Figure width must be greater than or equal to 600",
  "objectType": "MPCountValidationResult",
  "passed": false,
  "severity": 0,
  "type": "figure-minimum-width-resolution",
}
`;

exports[`validate manuscript validates a manuscript with a test template: validate-manuscript 25`] = `
Object {
  "_id": Any<String>,
  "data": Object {
    "count": 304,
    "id": "MPFigure:C9A3CBCE-4AF4-4AD2-97B0-D2848BEFDE5A",
    "value": 4000,
  },
  "objectType": "MPCountValidationResult",
  "passed": true,
  "severity": 0,
  "type": "figure-maximum-width-resolution",
}
`;

exports[`validate manuscript validates a manuscript with a test template: validate-manuscript 26`] = `
Object {
  "_id": Any<String>,
  "data": Object {
    "count": 350,
    "id": "MPFigure:C9A3CBCE-4AF4-4AD2-97B0-D2848BEFDE5A",
    "value": 400,
  },
  "message": "Figure height must be greater than or equal to 400",
  "objectType": "MPCountValidationResult",
  "passed": false,
  "severity": 0,
  "type": "figure-minimum-height-resolution",
}
`;

exports[`validate manuscript validates a manuscript with a test template: validate-manuscript 27`] = `
Object {
  "_id": Any<String>,
  "data": Object {
    "count": 350,
    "id": "MPFigure:C9A3CBCE-4AF4-4AD2-97B0-D2848BEFDE5A",
    "value": 1000,
  },
  "objectType": "MPCountValidationResult",
  "passed": true,
  "severity": 0,
  "type": "figure-maximum-height-resolution",
}
`;

exports[`validate manuscript validates a manuscript with a test template: validate-manuscript 28`] = `
Object {
  "_id": Any<String>,
  "data": Object {
    "count": 0,
    "value": 1,
  },
  "objectType": "MPCountValidationResult",
  "passed": true,
  "severity": 0,
  "type": "manuscript-maximum-corresponding-authors",
}
`;
